driver/src/driver/scheduler.rs

```diff
@@ -0,0 +1,274 @@
+        Ok(Self(time_since_epoch.as_secs() / BATCH_DURATION.as_secs()))
+    }
+
+    pub fn currently_being_solved(now: SystemTime) -> std::result::Result<Self, SystemTimeError> {
+        Self::current(now).map(|batch_id| batch_id.prev())
+    }
+
+    pub fn order_collection_start_time(self) -> SystemTime {
+        SystemTime::UNIX_EPOCH + Duration::from_secs(self.0 * BATCH_DURATION.as_secs())
+    }
+
+    pub fn solve_start_time(self) -> SystemTime {
+        self.order_collection_start_time() + BATCH_DURATION
+    }
+
+    pub fn next(self) -> BatchId {
+        self.0.checked_add(1).map(BatchId).unwrap()
+    }
+
+    pub fn prev(self) -> BatchId {
+        self.0.checked_sub(1).map(BatchId).unwrap()
+    }
+}
+
+#[derive(Debug)]
+pub struct AuctionTimingConfiguration {
+    /// The offset from the start of a batch at which point we should start
+    /// solving.
+    pub target_start_solve_time: Duration,
+    /// The offset from the start of the batch at which point there is not
+    /// enough time left to attempt to solve.
+    pub latest_solve_attempt_time: Duration,
+}
+
+pub struct Scheduler<'a> {
+    driver: &'a mut dyn StableXDriver,
+    auction_timing_configuration: AuctionTimingConfiguration,
+    last_solved_batch: Option<BatchId>,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+enum Action {
+    Solve(BatchId),
+    Sleep(Duration),
+}
+
+impl<'a> Scheduler<'a> {
+    pub fn new(
+        driver: &'a mut dyn StableXDriver,
+        auction_timing_configuration: AuctionTimingConfiguration,
+    ) -> Self {
+        assert!(auction_timing_configuration.latest_solve_attempt_time <= BATCH_DURATION);
+        assert!(
+            auction_timing_configuration.target_start_solve_time
+                < auction_timing_configuration.latest_solve_attempt_time
+        );
+        Self {
+            driver,
+            auction_timing_configuration,
+            last_solved_batch: None,
+        }
+    }
+
+    pub fn run_forever(&mut self) {
+        loop {
+            let now = SystemTime::now();
+            match self.determine_action(now) {
+                Ok(Action::Sleep(duration)) => {
+                    info!("Sleeping {}s.", duration.as_secs());
+                    thread::sleep(duration);
+                }
+                Ok(Action::Solve(batch_id)) => {
+                    info!("Solving batch {}.", batch_id.0);
+                    self.run_solver(batch_id);
+                }
+                Err(err) => {
+                    error!("Scheduler error: {}", err);
+                    thread::sleep(Duration::from_secs(10));
+                }
+            }
+        }
+    }
+
+    fn run_solver(&mut self, batch_id: BatchId) {
+        info!("Running solver for batch {}.", batch_id.0);
+        if let Err(err) = self.driver.run(batch_id.0.into()) {
+            error!("StableXDriver error: {}", err);
+            thread::sleep(Duration::from_secs(10));
```

Comment:
I would rather all sleeps be in the `run_forever` loop. I'm not sure what the best way to do this would be though.

Reply:
This one is just so we don't busy loop too hard in case we would retry. Imo it's fine and its not hidden as its just a few lines below where the function is called.

Reply:
The motivation was more about keeping all sleeping logic in the main looping function. If we move sleeps to other functions they have side-effects that aren't immediately obvious to someone who is reading `run_forever` for example.

Reply:
One way to model this is that the `match` expression in `run_forever` evaluates to a `Option<Duration>` and then `thread::sleep` is called with that duration if it is `Some`. That way we only have one sleep and the various match branches sleep with different values.

If we do it this way we an even refactor out a single `run_forever` iteration into its own method and have unit test(s) to verify that the duration that is supposed to be slept is as intended.

Reply:
Implemented this and added a test for it.

Reply:
Just reviewed it! Looks good :)